src/server/failover.c

```diff
@@ -806,6 +810,132 @@ alt_conn(pbs_net_t addr, unsigned int sec)
 
 /**
  * @brief
+ * 	Function to check if stonith script exists at PBS_HOME/server_priv/stonith.
+ * 	If it does then invoke the script for execution.
+ * 
+ * @param[in]	node - hostname of the node, that needs to be powered off.
+ * 
+ * @return	Error code
+ * @retval	 0 - stonith script executed successfully or script does not exist.
+ * @retval      -1 - stonith script failed to bring down node.
+ *
+ */
+int
+check_and_invoke_stonith(char *node)
+{
+	char		stonith_cmd[3*MAXPATHLEN+1] = {0};
+	char		stonith_fl[MAXPATHLEN+1] = {0};
+	char		*p = NULL;
+	char		out_err_fl[MAXPATHLEN+1] = {0};
+	char		*out_err_msg = NULL;
+	int		rc = 0;
+	int		fd = 0;
+	struct stat	stbuf;
+	
+	if (node == NULL )
+		return -1;
+
+	snprintf(stonith_fl, sizeof(stonith_fl), "%s/server_priv/stonith", pbs_conf.pbs_home_path);
+	stonith_fl[sizeof(stonith_fl) - 1] = '\0';
+
+#ifdef WIN32
+	repl_slash(stonith_fl);
+#endif /* WIN32 */
+	if (stat(stonith_fl, &stbuf) != 0) {
+		if (errno == ENOENT) {
+			snprintf(log_buffer, LOG_BUF_SIZE, "Skipping STONITH");
+			log_buffer[LOG_BUF_SIZE - 1] = '\0';
+			log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO,
+				  msg_daemonname, log_buffer);
+			return 0;
+		}
+	}
+
+	/* create unique filename by appending pid */
+	snprintf(out_err_fl, sizeof(out_err_fl), "%s/spool/stonith_out_err_fl_%s_%d", pbs_conf.pbs_home_path, node, getpid());
+	out_err_fl[sizeof(out_err_fl) - 1] = '\0';
+#ifdef WIN32
+	repl_slash(out_err_fl);
+
+	/* execute stonith script and redirect output to file */
```

Comment:
Looks like we still have some redundant code for both platforms which can be optimized.

Reply:
Have updated it.